tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,429 @@
+        X509_STORE_CTX_cleanup(ctx);
+    }
+    s2n_stuffer_free(&cert_chain_in_stuffer);
+    return err_code;
+}
+
+
+s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(struct s2n_x509_validator *validator,
+                                                                                const uint8_t *ocsp_response_raw,
+                                                                                uint32_t ocsp_response_length,
+                                                                                struct s2n_config *config) {
+    if (!validator->validate_certificates || !validator->check_stapled_ocsp) {
+        return S2N_CERT_OK;
+    }
+
+    OCSP_RESPONSE *ocsp_response = NULL;
+    OCSP_BASICRESP *basic_response = NULL;
+
+    s2n_cert_validation_code ret_val = S2N_CERT_OK;
+
+    if (!ocsp_response_raw) {
+        return S2N_CERT_ERR_INVALID;
+    }
+
+    ocsp_response = d2i_OCSP_RESPONSE(NULL, &ocsp_response_raw, ocsp_response_length);
+
+    if (!ocsp_response) {
+        ret_val = S2N_CERT_ERR_INVALID;
+        goto clean_up;
+    }
+
+    int ocsp_status = OCSP_response_status(ocsp_response);
+
+    if (ocsp_status != OCSP_RESPONSE_STATUS_SUCCESSFUL) {
+        ret_val = S2N_CERT_ERR_INVALID;
+        goto clean_up;
+    }
+
+    basic_response = OCSP_response_get1_basic(ocsp_response);
+    if (!basic_response) {
+        ret_val = S2N_CERT_ERR_INVALID;
+        goto clean_up;
+    }
+
+    int i;
+
+    int certs_in_chain = sk_X509_num(validator->cert_chain);
+    int certs_in_ocsp = sk_X509_num(basic_response->certs);
+
+    if (certs_in_chain >= 2 && certs_in_ocsp >= 1) {
+        X509 *responder = sk_X509_value(basic_response->certs, certs_in_ocsp - 1);
+
+        //check to see if one of the certs in the chain is an issuer of the cert in the ocsp response.
+        //if so it needs to be added to the OCSP verification chain.
+        for (i = 0; i < certs_in_chain; i++) {
+            X509 *issuer = sk_X509_value(validator->cert_chain, i);
+            if (X509_check_issued(issuer, responder) == X509_V_OK) {
+                if (!OCSP_basic_add1_cert(basic_response, issuer)) {
+                    ret_val = S2N_CERT_ERR_INVALID;
+                    goto clean_up;
+                }
+            }
+        }
+    }
+
+    //do the crypto checks on the response.
+    if (!OCSP_basic_verify(basic_response, validator->cert_chain, validator->trust_store->trust_store, 0)) {
+        ret_val = S2N_CERT_ERR_EXPIRED;
+        goto clean_up;
+    }
+
+    //for each response check the timestamps and the status.
+    for (i = 0; i < OCSP_resp_count(basic_response); i++) {
+        int status_reason;
+        ASN1_GENERALIZEDTIME *revtime, *thisupd, *nextupd;
+
+        OCSP_SINGLERESP *single_response = OCSP_resp_get0(basic_response, i);
+        if (!single_response)
+            continue;
+
+        ocsp_status = OCSP_single_get0_status(single_response, &status_reason, &revtime,
```

Comment:
`revtime` is never checked

Reply:
V_OCSP_CERTSTATUS_REVOKED will be returned if revtime has been set, it's only there for the function call, but we don't need it since OCSP_single_get0_status will fail if the certificate has been revoked.

Reply:
unless theres something else you think I should be doing with the revocation time.